epoxy-paging3/src/main/java/com/airbnb/epoxy/paging3/PagedDataModelCache.kt

```diff
@@ -0,0 +1,179 @@
+ * models for each item when they are invalidated to avoid rebuilding models for the whole list when pagingData is
+ * updated.
+ *
+ * The pagingData submitted to this cache must be kept in sync with the model cache. To do this,
+ * the executor of the pagingData differ is set to the same thread as the model building handler.
+ * However, change notifications from the PageList happen on that list's notify executor which is
+ * out of our control, and we require the user to configure that properly, or an error is thrown.
+ *
+ * There are two special cases:
+ *
+ * 1. The first time models are built happens synchronously for immediate UI. In this case we don't
+ * use the model cache (to avoid data synchronization issues), but attempt to fill the cache with
+ * the models later.
+ *
+ * 2. When a pagingData is submitted it can trigger update callbacks synchronously. Since we don't control
+ * that thread we allow a special case of cache modification when a new list is being submitted,
+ * and all cache access is marked with @Synchronize to ensure safety when this happens.
+ */
+internal class PagedDataModelCache<T : Any>(
+    private val modelBuilder: (itemIndex: Int, item: T?) -> EpoxyModel<*>,
+    private val rebuildCallback: () -> Unit,
+    itemDiffCallback: DiffUtil.ItemCallback<T>,
+    private val modelBuildingHandler: Handler
+) {
+    /**
+     * Backing list for built models. This is a full array list that has null items for not yet build models.
+     */
+    private val modelCache = arrayListOf<EpoxyModel<*>?>()
+
+    /**
+     * Tracks the last accessed position so that we can report it back to the paged list when models are built.
+     */
+    private var lastPosition: Int? = null
+
+    /**
+     * Set to true while a new list is being submitted, so that we can ignore the update callback
+     * thread restriction.
+     */
+    private var inSubmitList: Boolean = false
+
+    /**
+     * Observer for the PagedList changes that invalidates the model cache when data is updated.
+     */
+    private val updateCallback = object : ListUpdateCallback {
+        override fun onChanged(position: Int, count: Int, payload: Any?) = synchronizedWithCache {
+            assertUpdateCallbacksAllowed()
+            (position until (position + count)).forEach {
+                modelCache[it] = null
+            }
+            rebuildCallback()
+        }
+
+        override fun onMoved(fromPosition: Int, toPosition: Int) = synchronizedWithCache {
+            assertUpdateCallbacksAllowed()
+            val model = modelCache.removeAt(fromPosition)
+            modelCache.add(toPosition, model)
+            rebuildCallback()
+        }
+
+        override fun onInserted(position: Int, count: Int) = synchronizedWithCache {
+            assertUpdateCallbacksAllowed()
+            repeat(count) {
+                modelCache.add(position, null)
+            }
+            rebuildCallback()
```

Comment:
I faced with an [issue](https://issuetracker.google.com/issues/175033266) in paging. ListUpdateCallback methods are called before data actually changed. So we will get exception here:
```kotlin
java.lang.IndexOutOfBoundsException: Illegal attempt to access index 0 in ItemSnapshotList of size 0
        at androidx.paging.ItemSnapshotList.get(ItemSnapshotList.kt:70)
        at com.airbnb.epoxy.paging3.PagedDataModelCache.getModels(PagedDataModelCache.kt:145)
        at com.airbnb.epoxy.paging3.PagingDataEpoxyController.buildModels(PagingDataEpoxyController.kt:47)
        at com.airbnb.epoxy.EpoxyController$1.run(EpoxyController.java:281)
        at com.airbnb.epoxy.EpoxyController.requestModelBuild(EpoxyController.java:166)
        at com.airbnb.epoxy.paging3.PagingDataEpoxyController$modelCache$2.invoke(PagingDataEpoxyController.kt:40)
        at com.airbnb.epoxy.paging3.PagingDataEpoxyController$modelCache$2.invoke(PagingDataEpoxyController.kt:12)
        at com.airbnb.epoxy.paging3.PagedDataModelCache$updateCallback$1$onInserted$1.invoke(PagedDataModelCache.kt:78)
        at com.airbnb.epoxy.paging3.PagedDataModelCache$updateCallback$1$onInserted$1.invoke(PagedDataModelCache.kt:57)
        at com.airbnb.epoxy.paging3.PagedDataModelCache$updateCallback$1.synchronizedWithCache(PagedDataModelCache.kt:91)
        at com.airbnb.epoxy.paging3.PagedDataModelCache$updateCallback$1.onInserted(PagedDataModelCache.kt:73)
        at androidx.paging.AsyncPagingDataDiffer$differCallback$1.onInserted(AsyncPagingDataDiffer.kt:53)
        at androidx.paging.AsyncPagingDataDiffer$differBase$1.presentNewList(AsyncPagingDataDiffer.kt:78)
        at androidx.paging.PagingDataDiffer$collectFrom$2$invokeSuspend$$inlined$collect$1$lambda$1.invokeSuspend(PagingDataDiffer.kt:129)
```

Maybe it's better to call `requestModelBuild()` after `asyncDiffer.submitData()` (line 126)?

Reply:
Hm, the issue was gone when I removed this line 
https://github.com/airbnb/epoxy/blob/4528a5b274642a82dde8a568c5a2c2fbcafbfe4d/epoxy-paging3/src/main/java/com/airbnb/epoxy/paging3/PagedDataModelCache.kt#L120

Reply:
🎉 The bug will be fixed in v3.0.0-alpha11

Reply:
@keramat-farid @osipxd are there any extra works needed to be done? I think it can be upgraded to version `v3.0.0-alpha12` now and be finalized? I would appreciate a lot if this get to production soon as my company is relying a lot on this feature. Thank you!

Reply:
Yes, it will be enough to fix the issue.